### ISSUE TITLE
mAP always 0

### ISSUE THREAD

**Issue by sam-va** [ultralytics/yolov5#9045]
### Search before asking

- [ ] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I Have been trying to custom train YOLOv5 model with one label and I have made suitable changes in YAML file .However, 
mAP value is still 0 no matter how many epochs I run the training for. I have been struggling with this issue for many days. 
Can someone suggest a solution?
![Screenshot 2022-08-20 141130](https://user-images.githubusercontent.com/81673432/185736742-ba94dfcc-4ee6-4d65-9ccc-f8e10f1b3e97.png)

**Comment by semihahishali** [comment]
It means that your labelled image number is insufficent or you have choosen the bigger Yolov5 weight files https://github.com/ultralytics/yolov5#pretrained-checkpoints, you should try smaller ones.  If you have enough images, you should wait for a while and see these numbers are getting higher

**Comment by triple-Mu** [comment]
> It means that your labelled image number is insufficent or you have choosen the bigger Yolov5 weight files https://github.com/ultralytics/yolov5#pretrained-checkpoints, you should try smaller ones.  If you have enough images, you should wait for a while and see these numbers are getting higher

Maybe that's not the case.
I test that when training model, the Conv model's bn has nan mean and var.
If I do something like print bn's mean and var, It runs correctly.

**Comment by glenn-jocher** [comment]
@sam-va @triple-Mu @semihahishali good news 😃! Your original zero-mAP issue may now be fixed ✅ in PR #9037. 

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by glenn-jocher** [comment]
@sam-va @triple-Mu @semihahishali good news 😃! Your original issue may now be fixed ✅ in PR #9068 by @0zppd. @pourmand1376 tracked down the problem using `git bisect` and running each training 10x since the bug was not reproducible on every training, but showed up in maybe 1/3 of the trainings. 

The original issue was that I had replaced torch.zeros() with torch.empty() on some ops like warmup and profiling to try to get some slight speed improvements, and once op in particular ran a torch.empty() tensor through the model when it was in .train() mode, leading the batchnorm layers to add those values to the tracked statistics. Since torch.empty is not initialized it can take on extremely high or low values, leading some batchnorm layers to randomly output NaN. 

The PR has been extensively tested on 10x Colab trainings and all 10 came back good now:
<img width="1211" alt="Screenshot 2022-08-21 at 15 40 00" src="https://user-images.githubusercontent.com/26833433/185793959-986ad082-c181-49b8-a564-b96270ebc409.png">

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by mr-mainak** [comment]
Having the same problem. I installed CUDA:11.3 with cuDNN:8.2.0 and Pytorch:1.12.1+cu113 on Ubuuntu:20.04. Any suggestion on what might be the reason.

**Comment by glenn-jocher** [comment]
@mr-mainak it seems like you have the correct versions of CUDA, cuDNN, PyTorch, and Ubuntu installed. However, there could be other factors contributing to the issue you're experiencing.

To help troubleshoot, please provide more information about the specific problem you're facing. Did you encounter any error messages or unexpected behavior? It would be helpful if you could share the relevant portion of your code or any error logs.

In the meantime, you can try the following troubleshooting steps:
1. Double-check that all dependencies and packages are installed correctly by following the installation instructions in the YOLOv5 repository.
2. Ensure that you have the required dataset and labeled images in the correct format for training.
3. Check if any specific modifications or customizations you made to the code might be causing the issue.
4. Consider trying different YOLOv5 model architectures or pretrained weights to see if the issue persists.

By providing more details, we'll be able to assist you further in resolving the problem.

**Comment by mr-mainak** [comment]
> @mr-mainak it seems like you have the correct versions of CUDA, cuDNN, PyTorch, and Ubuntu installed. However, there could be other factors contributing to the issue you're experiencing.
> 
> To help troubleshoot, please provide more information about the specific problem you're facing. Did you encounter any error messages or unexpected behavior? It would be helpful if you could share the relevant portion of your code or any error logs.
> 
> In the meantime, you can try the following troubleshooting steps:
> 
> 1. Double-check that all dependencies and packages are installed correctly by following the installation instructions in the YOLOv5 repository.
> 2. Ensure that you have the required dataset and labeled images in the correct format for training.
> 3. Check if any specific modifications or customizations you made to the code might be causing the issue.
> 4. Consider trying different YOLOv5 model architectures or pretrained weights to see if the issue persists.
> 
> By providing more details, we'll be able to assist you further in resolving the problem.

@glenn-jocher I just upgraded pytorch to 2.0 version and everything is working fine. However still not sure why there was a problem with the previous pytorch version.

**Comment by glenn-jocher** [comment]
@mr-mainak glad to hear that upgrading to PyTorch 2.0 resolved the issue for you! It's possible that there were compatibility or bug-related issues with the previous version of PyTorch that were causing the problem.

While it's difficult to pinpoint the exact cause without more information, it's not uncommon for different versions of libraries and frameworks to have compatibility issues or bug fixes that can affect the functionality of certain applications or models.

If you encounter any further issues or have any other questions, feel free to reach out. We're here to help!
